Add backup_config resource (#241)
cockroachdb/terraform-provider-cockroach: docs/resources/backup_config.md

---
# generated by https://github.com/hashicorp/terraform-plugin-docs
page_title: "cockroach_backup_config Resource - terraform-provider-cockroach"
subcategory: ""
description: |-
  The backup configuration for a cluster.
  Each cluster has a backup configuration that determines if backups are enabled,
  how frequently they are taken and how often they are retained for.  This

>   resource allows configuring those settings.  It's important to note that the

Author: rereading this, I realized, 

and how often they are retained for

should be:

and how long they are retained for

  existence or lack of this resource doesn't not indicate whether a backup
  configuration exists or not. Rather, if Terraform is managing it or not.
  As a result of this, removal of the resource from the Terraform configuration

>   will leave the backup configuration in the state it was before the removal.

Reviewer: Does this mean, if you remove the backup_config tf resource, it just won't have any affect on the the configuration at all? If I have understood that correctly, perhaps it could be something like:

> As a result, removing the resource from the Terraform configuration will not affect the settings in the backup configuration.

Author: yes that's right. The suggestion sounds good to me.

  retention_days Edit Restrictions

>   retention_days can only be set once.  Further changes require opening a support
>   ticket.  For this reason, managing the value via terraform can require special
>   care. Here are some useful tips:

Reviewer: ```suggestion
  `retention_days` can only be set once. It is necessary to open a support ticket to modify the settings again. For this reason, consider the following when managing the value in Terraform:
```

Author: Since this only applies to this one setting, I've taken the suggestion but updated it to:

modify the setting again.

>   * Optionally, you can refrain from including retention_days in the
>   terraform resource to fully rely on server side management of the value and not need any
>   special care.

Reviewer: ```suggestion
  * (Optional) Refrain from including retention_days in the
  Terraform resource to rely on server-side management of the value instead.
```

>   * If the initial value for retention_days in the backupconfig resource is the

Reviewer: ```suggestion
  * If the initial value for `retention_days` in the `cockroach_backup_config` resource is the
```

>   default value (30), the single edit will not be used up and it can be changed one
>   more time.

Reviewer: ```suggestion
  default value (30), it will be possible to modify the retention setting one more time.
```

>   * If the initial value set for retention_days in the backupconfig resource is

Reviewer: ```suggestion
  * If the initial value set for `retention_days` in the `cockroach_backup_config` resource is
```

>   not the default, the single edit will be used up on the first apply. Further

Reviewer: ```suggestion
  not the default, it will not be possible to modify the retention setting again. Further
```

>   edits will require a support ticket.

Reviewer: ```suggestion
  modifications will require a support ticket.
```

>   Changing the value of retention_days after using your one change will be a
>   multi-step operation. Here are two workflows that will work.  Both of these
>   paths assume you already have a backup config resource that doesn't have any
>   updates left:
>   Change it and then open a ticket before applying:
>   
>   Update retention_days to the new desired value in terraform.Before applying the run, contact support to change the retention_days to the desired value.Apply the changes in terraform, a terraform READ operation will be done and see the existing value and update the tf state.Temporarily remove management of it from tf, update it via ticket and then add it back.
>   
>   Remove management of retention_days from the terraform resourceRun the apply, nothing will change but terraform is no longer managing that value.Open a support ticket to update retention_days to the desired value.Optionally, add retention_days back to the terraform config with the new value and apply the no-op update.

Reviewer: ```suggestion
  Changing the value of retention_days after using your one change will be a
  multi-step operation. Refer to the two following workflows. Both of these
  paths assume you already have a cockroach_backup_config resource that cannot be modified again. Change it and then open a ticket before applying:
  
  Update retention_days to the new value in Terraform. Before applying the run, contact support to change the retention_days to the new value. Apply the changes in Terraform, a Terraform READ operation will complete, recognize the existing value, and update the tf state. Temporarily remove management of it from tf, update it via ticket, and then add it back.
  
  Remove management of retention_days from the Terraform resource. Run the apply, nothing will change but Terraform is no longer managing that value. Open a support ticket to update retention_days to the new value. Optionally, add retention_days back to the Terraform config with the new value and apply the no-op update.
```

Reviewer: Ah I see the duplicate content in the lower part of this file that is formatted in lists, I've added the language updates here to the latter part of the file.

Author: yes this is a bit confusing.  This md file is generated from the content in the resource.  The description in the frontmatter looks to be a stripped down version of the contents of the file.

---

# cockroach_backup_config (Resource)

> The backup configuration for a cluster.

Reviewer: ```suggestion
The backup settings for a cluster.
```

In the docs, we are using "backup settings". I don't know if we need to keep this consistent, so this is a total nit.

	1. Update `retention_days` to the new desired value in terraform.
	1. Before applying the run, contact support to change the `retention_days` to the desired value.
	1. Apply the changes in terraform, a terraform READ operation will be done and see the existing value and update the tf state.
* Temporarily remove management of it from tf, update it via ticket and then add it back.
	1. Remove management of `retention_days` from the terraform resource
	1. Run the apply, nothing will change but terraform is no longer managing that value.
	1. Open a support ticket to update `retention_days` to the desired value.
	1. Optionally, add `retention_days` back to the terraform config with the new value and apply the no-op update.

## Example Usage

```terraform
resource "cockroach_backup_config" "backed_up_cluster" {
  id                = cockroach_cluster.backed_up_cluster.id
  enabled           = true
  frequence_minutes = 60
  retention_days    = 30
}

resource "cockroach_backup_config" "no_backups_cluster" {
  id      = cockroach_cluster.no_backups_cluster.id
  enabled = false
}
```

<!-- schema generated by tfplugindocs -->
## Schema

### Required

> - `enabled` (Boolean) Indicates whether the backup configuration is enabled. If set to false, no backups will be created.

Author: yes it indicates if backups are enabled.  Is the edit here:

`Indicates whether backups are enabled.` ?

> - `enabled` (Boolean) Indicates whether the backup configuration is enabled. If set to false, no backups will be created.

Reviewer 2: Drive by to say yes I think that's what she means, to remove the indirection of "the configuration" being enabled vs backups being enabled.
